### PR TITLE
Formula renames improvements

### PR DESCRIPTION
**Pull request by vladshablinsky**
* Add `--force` to skip check whether installed and migrating formulae are from the same taps.
* Add "Renamed Formulae" block to report.

@xu-cheng @mikemcquaid this is a branch for improvements for https://github.com/Homebrew/homebrew/pull/41006.

I'm going to work on improvements on this branch until everybody is happy.

### CODE REVIEW

File: Library/Homebrew/migrator.rb [hunk 1]
```diff
@@ -55,7 +55,10 @@ def initialize(formula)
 
     @old_tabs = oldpath.subdirs.each.map { |d| Tab.for_keg(Keg.new(d)) }
     @old_tap = old_tabs.first.tap
-    raise MigratorDifferentTapsError.new(formula, old_tap) unless from_same_taps?
+
+    unless ARGV.force? || from_same_taps?
```
Comment (mikemcquaid) [inline Library/Homebrew/migrator.rb]: Mind using an `if` here instead; I always find `unless` with multiple arguments hard to parse mentally.

File: Library/Homebrew/cmd/update.rb [hunk 2]
```diff
@@ -319,18 +319,51 @@ def empty?
...
+        if newpath
+          "#{tap}/#{path.basename(".rb")} -> #{tap}/#{newpath.basename(".rb")}"
+        else
+          "#{tap}/#{path.basename(".rb")}"
+        end
+      elsif newpath
+        "#{path.basename(".rb")} -> #{newpath.basename(".rb")}"
       else
-        path.basename(".rb").to_s
+        "#{path.basename(".rb")}"
```
Comment (mikemcquaid) [inline Library/Homebrew/cmd/update.rb]: Can leave this line as it was (`path.basename(".rb").to_s`)

File: Library/Homebrew/cmd/update.rb [hunk 3]
```diff
@@ -319,18 +319,51 @@ def empty?
...
+    @hash[:A] -= @hash[:R].map(&:last) if @hash[:A]
+    @hash[:D] -= @hash[:R].map(&:first) if @hash[:D]
+  end
+
+  def select_formula(key)
+    fetch(key, []).map do |path, newpath|
+      if path.to_s =~ HOMEBREW_TAP_PATH_REGEX
+        tap = "#{$1}/#{$2.sub("homebrew-", "")}"
+        if newpath
+          "#{tap}/#{path.basename(".rb")} -> #{tap}/#{newpath.basename(".rb")}"
```
Comment (xu-cheng) [inline Library/Homebrew/cmd/update.rb]: I found it's strange to have a method named `select_formula` returning string which is not a formula name. I suggest you create a new method to do this.

**Comment by xu-cheng** [comment]
You should also update https://github.com/Homebrew/homebrew/blob/master/Library/Homebrew/cmd/update.rb#L63-L91

It would be better if you can avoid duplicated computation.

File: Library/Homebrew/cmd/update.rb (hunk 3 above)
Comment (vladshablinsky) [inline Library/Homebrew/cmd/update.rb]: @xu-cheng what if I make `select_formula` return `[oldname, newname]` for `:R` key and we will be able to use `select_formulae(:R)` when performing migrations.`select_formula` will still return strings that are formula names. For that we need to update `report` after all the pulls from remote repositories. For `dump_formula_report` we can check if key is `:R`.

**Comment by vladshablinsky** [comment]
@xu-cheng I updated `select_formula` as I said in the [comment](https://github.com/Homebrew/homebrew/pull/42801#discussion_r36774696)

Now there is no duplicated computation when we perform migrations. `select_formula` now returns formula names, do we still need to add the method for renamed formulae?

File: Library/Homebrew/migrator.rb [hunk 4]
```diff
@@ -55,7 +55,10 @@ def initialize(formula)
 
     @old_tabs = oldpath.subdirs.each.map { |d| Tab.for_keg(Keg.new(d)) }
     @old_tap = old_tabs.first.tap
-    raise MigratorDifferentTapsError.new(formula, old_tap) unless from_same_taps?
+
+    if !ARGV.force? && !from_same_taps?
+      raise MigratorDifferentTapsError.new(formula, old_tap)
```
Comment (xu-cheng) [inline Library/Homebrew/migrator.rb]: I think you need to update the message inside `MigratorDifferentTapsError`.

Comment (vladshablinsky) [inline Library/Homebrew/migrator.rb]: @xu-cheng updated.

File: Library/Homebrew/migrator.rb [hunk 5]
```diff
@@ -42,6 +42,8 @@ def initialize(formula, tap)
         #{formula.name} from #{formula.tap} is given, but old name #{formula.oldname} was installed from #{tap}
 
         Please try to use #{name} to refer the formula
+
+        To force migrate use `brew migrate --force #{formula.oldname}`
```
Comment (xu-cheng) [inline Library/Homebrew/migrator.rb]: The output may need some improvement. It looks strange to have these empty lines. Maybe it's a good idea to eliminate the if-else branch as well.

In the mean time, the manpage need update as well.
